static-build/posts/2021/09/cors/index.md

```diff
@@ -0,0 +1,404 @@
+
+Although the `Origin` request header doesn't result in a different response body, it does change the response headers (`Access-Control-Allow-Origin` in this case), so it needs to go in `Vary`.
+
+# Unusual requests and preflights
+
+So far, the response has been opting into exposing its data. All of the _requests_ have been assumed to be safe, because they're not doing anything unusual.
+
+```js
+fetch(url, { credentials: 'include' });
+```
+
+There's nothing unusual about the above, because the request is really similar to what an `<img>` can do already.
+
+```js
+fetch(url, {
+  method: 'POST',
+  body: formData,
+});
+```
+
+There's nothing unusual about the above, because the request is really similar to what a `<form>` can already do.
+
+```js
+fetch(url, {
+  method: 'wibbley-wobbley',
+  credentials: 'include',
+  headers: {
+    fancy: 'headers',
+    'here-we': 'go',
+  },
+});
+```
+
+Ok, that's pretty unusual.
+
+What counts as 'unusual' is pretty complicated, but at a high level, if it's the kind of request that other browser APIs don't generally make, then it's unusual. At a lower level, if the request method isn't `GET`, `HEAD`, or `POST`, or it includes headers or header values that aren't part of the [safelist](https://fetch.spec.whatwg.org/#cors-safelisted-request-header), then it counts as unusual.
+
+If you try to make an unusual request, the browser first asks the other origin if it's ok to send it. This process is called a _preflight_.
+
+## Preflight request
+
+Before making the main request, the browser makes a preflight request to the destination URL with a method of `OPTIONS`, and headers like this:
+
+```
+Access-Control-Request-Method: wibbley-wobbley
+Access-Control-Request-Headers: fancy, here-we
+```
+
+- `Access-Control-Request-Method` - The HTTP method that the main request will use. This is included even if the method isn't unusual.
+- `Access-Control-Request-Headers` - The unusual headers that the main request will use. If there are no unusual headers, this header isn't sent.
+
+The preflight request never includes credentials, even if the main request will.
+
+## Preflight response
+
+The server responds to indicate whether it's happy for the main request to go ahead, using headers like this:
+
+```
+Access-Control-Allow-Methods: wibbley-wobbley
+Access-Control-Allow-Headers: fancy, here-we
+```
+
+- `Access-Control-Allow-Methods` - The _unusual_ methods to allow. This can be a comma-separated list. If the main request is to be sent without credentials, this can be `*` to allow any method.
+- `Access-Control-Allow-Headers` - The _unusual_ headers to allow. If the main request is to be sent without credentials, this can be `*` to allow any header that isn't on the [💀 FORBIDDEN LIST 💀](https://fetch.spec.whatwg.org/#forbidden-header-name).
```

Comment:
If you could add something about why `*` doesn't allow these 'forbidden' headers, that could be helpful! From a first glance at the list it's not exactly clear why these headers aren't included in `*`. And is one of the headers on the forbidden list valid on its own? eg. `Access-Control-Allow-Headers: Accept-Encoding`? Or does it just apply to using `*`.